File: ambeth/Ambeth.Util/ambeth/collections/SetLinkedIterator.cs
using De.Osthus.Ambeth.Util;
using System;
using System.Text;

namespace De.Osthus.Ambeth.Collections
{
    public class SetLinkedIterator<E, K> : AbstractIterator<K> where E : SetLinkedEntry<K>
    {
        protected SetLinkedEntry<K> currEntry;

        protected IListElem<SetLinkedEntry<K>> next;

        protected bool first = true;

        private readonly AbstractLinkedSet<E, K> hashSet;

        private readonly InterfaceFastList<E> fastIterationList;

        public SetLinkedIterator(AbstractLinkedSet<E, K> hashSet, InterfaceFastList<E> fastIterationList, bool removeAllowed)
            : base(removeAllowed)
        {
            this.hashSet = hashSet;
            this.fastIterationList = fastIterationList;
        }

        public override void Dispose()
        {
            // Intended blank
        }

        public override bool MoveNext()
        {
            if (first)
            {
                IListElem<E> firstElem = fastIterationList.First;
                if (firstElem == null)
                {
                    return false;
                }
                currEntry = firstElem.ElemValue;
                next = currEntry.Next;
                first = false;
                return true;
            }
            else if (next == null)
            {
                return false;
            }
            currEntry = next.ElemValue;
            next = currEntry.Next;
            return true;
        }

        public override K Current
        {
            get
            {
                return currEntry.Key;
            }
        }

        public override void Remove()
        {
            if (!removeAllowed)
            {
                throw new NotSupportedException();
            }
            hashSet.Remove(currEntry.Key);
            currEntry = null;
        }
    }
}

File: jambeth/jambeth-util/src/main/java/de/osthus/ambeth/collections/SetLinkedIterator.java
package de.osthus.ambeth.collections;

public class SetLinkedIterator<K> extends AbstractIterator<K>
{
	protected SetLinkedEntry<K> currPointer, lastPointer;

	private final AbstractLinkedSet<K> hashSet;

	public SetLinkedIterator(final AbstractLinkedSet<K> hashSet, boolean removeAllowed)
	{
		super(removeAllowed);
		this.hashSet = hashSet;
		currPointer = hashSet.fastIterationList.getFirstElem();
		lastPointer = null;
	}

	@Override
	public final boolean hasNext()
	{
		return currPointer != null;
	}

	@Override
	public final K next()
	{
		final K elem = currPointer.key;
		lastPointer = currPointer;
		currPointer = currPointer.next;
		return elem;
	}

	@Override
	public void remove()
	{
		if (!removeAllowed)
		{
			throw new UnsupportedOperationException();
		}
		hashSet.remove(lastPointer.key);
		lastPointer = null;
	}
}
